### ISSUE TITLE
kubectl v1.13 fails to convert extensions/NetworkPolicy

### ISSUE THREAD

**Issue by jessesuen** [argoproj/argo-cd#1012]
Encountered the following issue with an application which had extensions/NetworkPolicy. Downgrading the kubectl from v1.13.1 to v1.12.4 fixed the issue.

```
$ argocd app get APPNAME
Name:               APPNAME
Server:             <REDACTED>
Namespace:          APPNAME
URL:                https://<REDACTED>/applications/APPNAME
Repo:               <REDACTED>
Target:             HEAD
Path:               .
Environment:        qal
Sync Policy:        <none>
Sync Status:        Unknown (0ecd09d)
Health Status:      Missing

CONDITION        MESSAGE
ComparisonError  kubectl convert is DEPRECATED and will be removed in a future version.
In order to convert, kubectl apply the object to the cluster, then kubectl get at the desired version.
error: unable to decode "/dev/shm/573599277": no kind "NetworkPolicy" is registered for version "networking.k8s.io/v1" in scheme "k8s.io/kubernetes/pkg/api/legacyscheme/scheme.go:29"


GROUP                         KIND              NAMESPACE    NAME                                                           STATUS     HEALTH
extensions                    NetworkPolicy     APPNAME      allow-egress-knative                                           OutOfSync  Missing
...
...
```

We should stop using `kubectl` convert

**Comment by jessesuen** [comment]
The following kustomize overlay can be used to workaround to downgrade the kubectl used in the application-controller:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: argocd-application-controller
spec:
  template:
    spec:
      volumes:
      - name: custom-tools
        emptyDir: {}
      initContainers:
      - name: download-tools
        image: alpine:3.8
        command: [sh, -c]
        args:
        - wget https://storage.googleapis.com/kubernetes-release/release/v1.12.4/bin/linux/amd64/kubectl &&
          chmod +x kubectl &&
          mv kubectl /custom-tools
        volumeMounts:
        - mountPath: /custom-tools
          name: custom-tools
        resources:
          requests:
            memory: 64M
            cpu: 0.02
      containers:
      - name: argocd-application-controller
        volumeMounts:
        - mountPath: /usr/local/bin/kubectl
          name: custom-tools
          subPath: kubectl
```

**Comment by jessesuen** [comment]
Given that `kubectl convert` is deprecated, and not completely functioning in v1.13, we need stop using `kubectl convert` when we convert an object from its preferred version, into the version defined in the git config.

The options I see are:
1. Stop trying to convert objects and start storing all versions in the cache. Go back to using disco.ServerResources() instead of disco.ServerPreferredResources() and store all versions of objects our cache.
Pros: lookups remain quick. guarantees to work since we have all versions of the object for lookup.
Cons: This will open more connections to API server. Will also use more memory in controller, though I think the memory increase in controller is negligible.

2. Find a proper way to convert objects using k8s library instead of `kubectl convert`
Pros: connections and API requests are limited
Cons: Hard to get this right. Also, eventually it seems this approach will stop working since K8s is going in a direction where core types may not be available in the local scheme.Scheme, and we may not be able to convert objects in the client.

3. Perform a GET against K8s API server using git config version, to perform the conversion. 
Pros: behavior is correct since K8s API server will do conversion for us
Cons: every call to ManagedResources may perform N queries, where N is the number of objects in the app which are defined in git using a non-preferred API version (e.g. extensions/v1beta1).

**Comment by jessesuen** [comment]
A partial fix is in place (downgrade kubectl to v0.12.4) which is able to perform convert on extensions/NetworkPolicy. Leaving this bug open for a more permanent fix.

**Comment by alexmt** [comment]
This should be fixed, better sooner than later.  We cannot stick to  v0.12.4 forever.
